ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -418,8 +445,8 @@ public void onCompletion(String result, Exception exception) {
         }
       }
 
-      if (exception != null) {
-        submitResponse(restRequest, restResponseChannel, null, exception);
+      if (response != null || exception != null) {
+        submitResponse(restRequest, restResponseChannel, response, exception);
```

Comment:
That would be a programming error I think. However, to answer your question, `AsyncRequestResponseHandler::submitResponse()` prefers exceptions over responses.